# **Tutorial:** Custom YOLO Object Detection with RoboFlow

In this tutorial, you'll learn how to create a custom YOLO object detection model using your own classes and annotations with RoboFlow.

### **Step 1: Install Required Dependencies**

In [1]:
%pip install --upgrade ultralytics
%pip install opencv-python
%pip install roboflow
%pip install matplotlib
#%pip install numpy==1.26.4 --force-reinstall


[notice] A new release of pip is available: 26.0 -> 26.0.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.

[notice] A new release of pip is available: 26.0 -> 26.0.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.
  Using cached requests_toolbelt-1.0.0-py2.py3-none-any.whl.metadata (14 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.9/49.9 MB 17.0 MB/s  0:00:03m0:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 16.3 MB/s  0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 18.7 MB/s  0:00:00 eta 0:00:01
Using cached requests_toolbelt-1.0.0-py2.py3-none-any.whl (54 kB)
  Attempting uninstall: idna━━╺━━━━━━━━━━━━━━━━━ 5/9 [opencv-python-headless]
    Found existing installation: idna 3.1190m━━━━━━━━━━━━━━━━━ 5/9 [opencv-python-headless]
    Uninstalling idna-3.11:╺━━━━━━━━━━━━━━━━━ 5/9 [opencv-python-headless]
      Suc

### **Step 2: Import Required Libraries**

In [3]:
import cv2
from roboflow import Roboflow
from ultralytics import YOLO
import matplotlib.pyplot as plt
import numpy as np
import os
from IPython.display import Image, display

### **Step 3: Set Up RoboFlow Account**

1. Go to RoboFlow and create a free account
2. Create a new project and select "Object Detection" as project type
3. Choose "YOLOv26" as the model format

### **Step 4: Upload and Annotate Your Images**

#### Using RoboFlow Web Interface:

1. Click "Upload" to add your images
2. Use the annotation tool to draw bounding boxes around objects
3. Assign the correct class labels to each bounding box
4. Apply preprocessing and augmentation as needed

### **Step 5. Download your Custom your Dataset**

1. In RoboFlow, go to your project dashboard
2. Click "Generate" to create a dataset version
3. Apply any desired augmentations
4. Click "Export" and select "YOLOv**x**" format (**x** the version you are using)
5. Copy the download code

In [4]:
!pip install roboflow

from roboflow import Roboflow
rf = Roboflow(api_key="gdWaMVat2ltEXWstB11u")
project = rf.workspace("proyects-ohv2m").project("dataset_calculator_id_card")
version = project.version(3)
dataset = version.download("yolo26")
                


[notice] A new release of pip is available: 26.0 -> 26.0.1
[notice] To update, run: pip install --upgrade pip
loading Roboflow workspace...
loading Roboflow project...



Extracting Dataset Version Zip to Dataset_calculator_ID_card-3 in yolo26:: 100%|██████████| 400/400 [00:00<00:00, 937.59it/s] 


### **Step 6: Train Your Custom YOLO Model**

In [5]:
# Load the YOLO model with custom configuration
model = YOLO('yolo26n.pt')  # Using nano version as base

# Train the model on your custom dataset
results = model.train(
    data=f'{dataset.location}/data.yaml',  # Path to dataset config
    epochs=20,  # Number of training epochs
    imgsz=640,  # Image size
    batch=16,   # Batch size
    name='custom_yolo_model'  # Name for the training run
)

Ultralytics 8.4.22 🚀 Python-3.11.2 torch-2.5.1+cu121 CUDA:0 (NVIDIA GeForce GTX 1650, 3904MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/media/aldrinchp/KAli-HDD/Subjects/NOVENO/Computer vision/Computer_Vision/UC.04 Object Detection/Object Detection/Dataset_calculator_ID_card-3/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=20, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolo26n.pt, momentum=0.937, mosaic=1.0, m

### **Step 7: Evaluate Your Model**

In [6]:
# Evaluate model performance
metrics = model.val()  # Validate the model
print(f"mAP50-95: {metrics.box.map}")
print(f"mAP50: {metrics.box.map50}")

# Show some validation results


Ultralytics 8.4.22 🚀 Python-3.11.2 torch-2.5.1+cu121 CUDA:0 (NVIDIA GeForce GTX 1650, 3904MiB)
YOLO26n summary (fused): 122 layers, 2,375,226 parameters, 0 gradients, 5.2 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 18.5±8.0 MB/s, size: 217.3 KB)
val: Scanning /media/aldrinchp/KAli-HDD/Subjects/NOVENO/Computer vision/Computer_Vision/UC.04 Object Detection/Object Detection/Dataset_calculator_ID_card-3/valid/labels.cache... 17 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 17/17 7.1Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 2/2 8.1s/it 16.2s.9s
                   all         17         18      0.987      0.958      0.995      0.995
               ID_card         12         12          1      0.916      0.995      0.995
            calculator          6          6      0.974          1      0.995      0.995
Speed: 279.7ms preprocess, 190.8ms inference, 0.0ms loss, 0.4ms postprocess per image
Resul

In [7]:
!yolo task=detect mode=val model=runs/detect/custom_yolo_model/weights/best.pt data="{dataset.location}"/data.yaml

Ultralytics 8.4.22 🚀 Python-3.11.2 torch-2.5.1+cu121 CUDA:0 (NVIDIA GeForce GTX 1650, 3904MiB)
YOLO26n summary (fused): 122 layers, 2,375,226 parameters, 0 gradients, 5.2 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 37.0±34.5 MB/s, size: 220.4 KB)
val: Scanning /media/aldrinchp/KAli-HDD/Subjects/NOVENO/Computer vision/Computer_Vision/UC.04 Object Detection/Object Detection/Dataset_calculator_ID_card-3/valid/labels.cache... 17 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 17/17 2.7Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 2/2 2.7s/it 5.5s6.7s
                   all         17         18      0.987      0.958      0.995      0.995
               ID_card         12         12          1      0.916      0.995      0.995
            calculator          6          6      0.974          1      0.995      0.995
Speed: 3.2ms preprocess, 216.4ms inference, 0.0ms loss, 0.1ms postprocess per image
Result

### **Step 8: Test Your Custom Model on an Image**

In [ ]:
# Load your custom trained model
custom_model = YOLO('/Users/eugenio/Documents/Notebooks_ArtificialIntelligence/runs/detect/custom_yolo_model/weights/best.pt')

# Test on a sample image
results = custom_model('/Users/eugenio/Documents/Notebooks_ArtificialIntelligence/5. Computer Vision/Object Detection/taza_test.jpg', conf=0.25)

# Display results
%matplotlib inline
for result in results:
    img = result.plot()  # Plot detections
    img_rgb = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    plt.figure(figsize=(12, 8))
    plt.imshow(img_rgb)
    plt.axis('off')
    plt.show()

### **Step 9: Test Your Custom Model on Video**

In [1]:
from pathlib import Path
import cv2
from ultralytics import YOLO

# 1) Ruta fija al modelo entrenado
model_path = Path("runs/detect/custom_yolo_model/weights/best.pt")
if not model_path.exists():
    raise FileNotFoundError(f"No se encontró el modelo en: {model_path}")

# 2) Cargar modelo
model = YOLO(str(model_path))
classNames = model.names

# 3) Abrir cámara (intenta 0, luego 1)
captura = None
for cam_idx in [0, 1]:
    test_cap = cv2.VideoCapture(cam_idx)
    if test_cap.isOpened():
        captura = test_cap
        print(f"Cámara abierta en índice {cam_idx}")
        break
    test_cap.release()

if captura is None:
    raise RuntimeError("No se pudo abrir ninguna cámara (índices 0 o 1).")

# 4) Parámetros (parecido al notebook 2, pero con el modelo custom)
imgsz = 416
conf = 0.30
print("Detección iniciada. Presiona 'q' para salir.")

while True:
    success, img = captura.read()
    if not success:
        print("No se pudo leer un frame de la cámara.")
        break

    results = model(img, stream=True, imgsz=imgsz, conf=conf, verbose=False)

    for r in results:
        boxes = r.boxes
        for box in boxes:
            cls = int(box.cls[0])
            class_name = classNames[cls]

            x1, y1, x2, y2 = box.xyxy[0]
            x1, y1, x2, y2 = int(x1), int(y1), int(x2), int(y2)

            cv2.rectangle(img, (x1, y1), (x2, y2), (255, 0, 255), 2)
            label = f"{class_name} {float(box.conf[0]):.2f}"
            cv2.putText(
                img,
                label,
                (x1, max(y1 - 10, 20)),
                cv2.FONT_HERSHEY_SIMPLEX,
                0.7,
                (255, 0, 0),
                2
            )

    cv2.imshow('Webcam', img)

    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

captura.release()
cv2.destroyAllWindows()

[ WARN:0@4.834] global cap_v4l.cpp:914 open VIDEOIO(V4L2:/dev/video0): can't open camera by index
[video4linux2,v4l2 @ 0x30a265c0] Not a video capture device.
[video4linux2,v4l2 @ 0x30a15b80] ioctl(VIDIOC_G_INPUT): Inappropriate ioctl for device
[ERROR:0@5.158] global obsensor_uvc_stream_channel.cpp:163 getStreamChannelGroup Camera index out of range


Cámara abierta en índice 1
Detección iniciada. Presiona 'q' para salir.


/media/aldrinchp/KAli-HDD/Subjects/NOVENO/Computer vision/.venv/lib/python3.11/site-packages/torch/cuda/__init__.py:129: UserWarning: CUDA initialization: CUDA unknown error - this may be due to an incorrectly set up environment, e.g. changing env variable CUDA_VISIBLE_DEVICES after program start. Setting the available devices to be zero. (Triggered internally at ../c10/cuda/CUDAFunctions.cpp:108.)
  return torch._C._cuda_getDeviceCount() > 0
QFontDatabase: Cannot find font directory /media/aldrinchp/KAli-HDD/Subjects/NOVENO/Computer vision/.venv/lib/python3.11/site-packages/cv2/qt/fonts.
Note that Qt no longer ships fonts. Deploy some (from https://dejavu-fonts.github.io/ for example) or switch to fontconfig.
QFontDatabase: Cannot find font directory /media/aldrinchp/KAli-HDD/Subjects/NOVENO/Computer vision/.venv/lib/python3.11/site-packages/cv2/qt/fonts.
Note that Qt no longer ships fonts. Deploy some (from https://dejavu-fonts.github.io/ for example) or switch to fontconfig.
QFontDa

## **Documentación del Proyecto **

### **1) Objetivo del proyecto**
Entrenar un detector de objetos personalizado con YOLO para reconocer dos clases:
- **ID_card**
- **calculator**

El modelo se entrenó y validó con un dataset exportado desde Roboflow en formato YOLO.

### **2) Detalles del dataset**
- **Nombre del dataset:** Dataset_calculator_ID_card (v3)
- **Fuente:** Roboflow Universe
- **Licencia:** CC BY 4.0
- **Total de imágenes:** 194
- **Clases:** 2 (`ID_card`, `calculator`)
- **Split del dataset:**
  - Train: 168 imágenes
  - Valid: 17 imágenes
  - Test: 9 imágenes

**Preprocesamiento aplicado en Roboflow:**
- Auto-orientación de imágenes (EXIF)

**Aumentaciones usadas para generar versiones del dataset:**
- Volteo horizontal (50%)
- Recorte aleatorio (0% a 20%)
- Rotación aleatoria (-15° a +15°)
- Ajuste de brillo (-15% a +15%)
- Ajuste de exposición (-10% a +10%)
- Ruido sal y pimienta (0.1% de píxeles)

### **3) Configuración de entrenamiento**
- **Framework:** Ultralytics YOLO (YOLO26n)
- **Modelo base:** `yolo26n.pt`
- **Tarea:** Detección de objetos
- **Épocas:** 20
- **Batch size:** 16
- **Tamaño de imagen (`imgsz`):** 640
- **Optimizador:** automático (configuración por defecto de Ultralytics)
- **Aceleración:** CUDA (GPU NVIDIA GeForce GTX 1650)

### **4) Rendimiento del modelo entrenado**
Resultados reportados en validación:
- **Precisión global (P):** 0.987
- **Recall global (R):** 0.958
- **mAP@50:** 0.995
- **mAP@50-95:** 0.995

Métricas por clase (validación):
- **ID_card:** P = 1.000, R = 0.916, mAP@50 = 0.995, mAP@50-95 = 0.995
- **calculator:** P = 0.974, R = 1.000, mAP@50 = 0.995, mAP@50-95 = 0.995

Métricas finales (epoch 20, registro de entrenamiento):
- **P:** 0.996
- **R:** 0.948
- **mAP@50:** 0.995
- **mAP@50-95:** 0.995

### **5) Prueba del modelo e implementación en tiempo real**
Se realizó inferencia sobre imágenes/video y también detección en cámara web en tiempo real, mostrando:
- Caja delimitadora por objeto detectado
- Etiqueta de clase (`ID_card` o `calculator`)
- Confianza de la predicción en pantalla

### **6) Conclusión**
El modelo entrenado logra un desempeño alto y estable para ambas clases del proyecto, con **mAP@50-95 de 0.995**. Se cumple el objetivo de detección personalizada y la implementación en tiempo real funciona correctamente para el escenario planteado.

### **7) Evidencia generada**
Durante el entrenamiento y validación se generaron artefactos de soporte (curvas, matriz de confusión, resultados y pesos del modelo) en la carpeta `runs/detect/custom_yolo_model`, útiles para auditoría y presentación de resultados.

## **Activity: Create Your Custom Detector**

1. **Create a RoboFlow account** and set up a new project
2. **Upload at least 20 images** of your custom objects (at least 2)
3. **Annotate all images** with bounding boxes and class labels
4. **Train a YOLO model** on your custom dataset
5. **Test your model** on new images and evaluate its performance
6. **Implement real-time detection** with your custom classes

## **Grading Rubric:**

| **Criteria** | **Points** | **Description** |
|-------------|-----------|----------------|
| **1. Dataset Creation** | 3 | Successfully creates and annotates a custom dataset with at least 25 images and multiple classes in RoboFlow. |
| **2. Model Training** | 3 | Properly trains a YOLOv8 model on the custom dataset with appropriate hyperparameters. |
| **3. Model Evaluation** | 2 | Evaluates model performance with appropriate metrics and provides analysis of results. |
| **4. Custom Detection** | 2 | Implements real-time detection with custom classes and displays results correctly. |
| **5. Documentation** | 1 | Provides clear documentation of the process, including dataset details and model performance. |

**Total: 10 points**

## **Tips for Success:**

- **Quality over quantity**: Well-annotated images are more important than many poorly annotated ones
- **Class balance**: Try to have a similar number of examples for each class
- **Variety**: Include different angles, lighting conditions, and backgrounds
- **Augmentation**: Use RoboFlow's augmentation features to expand your dataset

<p style="text-align: right; font-size:14px; color:gray;">
<b>Prepared by:</b><br>
Manuel Eugenio Morocho-Cayamcela
</p>